### PR TITLE
Use memory bounded buffer in executors

### LINKED ISSUE #7531: The buffer in executor should support memory bound
## Enhancement
When running MPP query, some non-memory intensive executors will still buffer data(like `ExchangeReceiver/UnionBlockInputStream/SharedQueryBlockInputStream/ExchangeSender`). Currently the buffer inside these executors are usually size bounded, since the memory usage varies a lot between different elements in buffer, it will be more memory friendly if the buffer inside the executors are memory bounded.

### PR DESCRIPTION
**Pull request by windtalker**
### What problem does this PR solve?

Issue Number: close #7531

Problem Summary:

### What is changed and how it works?
- Use memory bounded buffer in ExchangeReceiver/MPPTunnel/UnionBlockInputStream/SharedQueryBlockInputStream/SharedQueue
### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
In a 3 nodes TiFlash cluster, run `explain analyze select * from lineitem` on TPCH-100

| max_buffered_bytes_in_executor | Peak memory usages in one TiFlash node |
|---------|------------------------------------------|
| 0(unlimited) |  2.37 GiB  |
| 200 MB(default value) |   1.55 GiB    |
| 1 byte | 1.39 GiB |

The query time is almost the same as it is a really simple query
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Flash/Mpp/ExchangeReceiver.h [hunk 1]
```diff
@@ -26,6 +26,8 @@
 #include <mutex>
 #include <thread>
 
+#include "Interpreters/Settings.h"
```
Comment (SeaRise) [inline dbms/src/Flash/Mpp/ExchangeReceiver.h:29]:
```suggestion
#include <Interpreters/Settings.h>
```

Comment (SeaRise) [reply dbms/src/Flash/Mpp/ExchangeReceiver.h:29]: Or just `class Settings;` here

File: dbms/src/Flash/Mpp/MPPTask.cpp [hunk 2]
```diff
@@ -174,6 +175,8 @@ void MPPTask::registerTunnels(const mpp::DispatchTaskRequest & task_request)
     auto tunnel_set_local = std::make_shared<MPPTunnelSet>(log->identifier());
     std::chrono::seconds timeout(task_request.timeout());
     const auto & exchange_sender = dag_context->dag_request.rootExecutor().exchange_sender();
+    size_t tunnel_queue_memory_bound = getAverageThreshold(context->getSettingsRef().max_buffered_bytes_in_executor, exchange_sender.encoded_task_meta_size());
```
Comment (SeaRise) [inline dbms/src/Flash/Mpp/MPPTask.cpp:178]: Some worry that memory_buffer is not enough in large-scale clusters.
For a scale of 100 nodes, a default queue is only 1mb

Comment (SeaRise) [reply dbms/src/Flash/Mpp/MPPTask.cpp:178]: ok

**Review by yibin87 — APPROVED** [review]
LGTM

**Pull request merged**